### DISCUSSION TITLE
Full Disk Encryption (FDE) with LUKS and GRUB: one way to speed it up

### DISCUSSION THREAD

**Discussion by marwenius** [SpiralLinux/SpiralLinux-project#497] [General]
When I tried SpiralLinux with FDE for the first time, [it took a very long time until the LUKS slots were opened after entering the password](https://github.com/SpiralLinux/SpiralLinux-project/issues/328).

It turns out that GRUB does not use hardware acceleration for decryption and is therefore slow.

I have solved it for myself by setting the number of iterations manually.

I use this command (from a live system):
```
sudo cryptsetup luksChangeKey --pbkdf-force-iterations 600000 /dev/sda2 --key-slot 0 --pbkdf pbkdf2
```
(sda2 must be adjusted accordingly)

If another encrypted partition is used for hibernation, the iterations should also be adjusted there.

The figure of 600,000 is just one example. I have based this on the iterations of BitWarden, which also work with 600,000, and also on other articles that I link to in the resources. For my purposes, I use a different number, but 600,000 might be a good start for some. Instead of waiting 30 seconds, I now only wait 3 to 4 seconds, which is acceptable for me.

In this context, I also convert directly from LUKS1 to LUKS2 (before I change the iterations). On the one hand because of better protection against header corruption and the option to switch from PBKDF2 to Argon2 if required:
```
sudo cryptsetup convert /dev/sda2 --type luks2
```

Unfortunately, Debian's version of GRUB currently only supports LUKS with PBKDF2 and no Argon2, which provides better protection against GPU-based attacks (although of course the question is whether your own threat level needs this extra protection). If one does not depend on Debian: There are Argon2-GRUB patches for Arch and Gentoo.

Resources:
- [Raising PBKDF2 iterations, password strength and StatiCrypt security model](https://github.com/robinmoisson/staticrypt/issues/159)
- [Full encryption with LUKS (SHA512, aes-xts-plain64) – grub2 really slow ….](https://linux-blog.anracom.com/2018/11/30/full-encryption-with-luks-sha512-aes-xts-plain64-grub2-really-slow/)
- [PSA: upgrade your LUKS key derivation function](https://mjg59.dreamwidth.org/66429.html) ([Discussion about this on Reddit](https://www.reddit.com/r/linux/comments/12q51ce/psa_upgrade_your_luks_key_derivation_function/))
- [GRUB/Tips and tricks: Speeding up LUKS decryption in GRUB](https://wiki.archlinux.org/title/GRUB/Tips_and_tricks#Speeding_up_LUKS_decryption_in_GRUB)
- [PBKDF2 vs Argon2 - Finally some hard numbers](https://www.reddit.com/r/Bitwarden/comments/1167rwm/pbkdf2_vs_argon2_finally_some_hard_numbers/)
- [PBKDF2 default now apparently 600,000 (for new accounts)](https://www.reddit.com/r/Bitwarden/comments/10kg8eu/pbkdf2_default_now_apparently_600000_for_new/)
- [Enable LUKS2 and Argon2 Support for Packages](https://leo3418.github.io/collections/gentoo-config-luks2-grub-systemd/packages.html)
- [Enable LUKS2 and Argon2 support for Grub in Manjaro/Arch](https://mdleom.com/blog/2022/11/27/grub-luks2-argon2/)
- Wikipedia: [Argon2](https://en.wikipedia.org/wiki/Argon2), [PBKDF2](https://en.wikipedia.org/wiki/PBKDF2), [LUKS](https://en.wikipedia.org/wiki/Linux_Unified_Key_Setup)

**Comment by azzamsa** [comment]
> sudo cryptsetup luksChangeKey --pbkdf-force-iterations 600000 /dev/sda2 --key-slot 0 --pbkdf pbkdf2

I am surprised because it ask for new password, turns out after some reading, it can't do the operation without changing the pasword.
*After reboot, it goes from ~1-2 minutes to 4.40s, as I expected.*
The default is 7,000,000.

Thanks a ton for this, I thought the only solution is fresh install and having separate unencrypted partition for boot.
Again, thank you so much, I was so close to do fresh install using different approach.
